File: packages/runtime-web-three/src/meshBounds.ts
import type { IAssetIr } from "@threenative/ir";

export type Vec3Tuple = readonly [number, number, number];

export interface IAabb {
  max: Vec3Tuple;
  min: Vec3Tuple;
}

export interface IBoundingSphere {
  center: Vec3Tuple;
  radius: number;
}

export function sampleMeshPoints(asset: IAssetIr, options: { maxSamples?: number } = {}): Vec3Tuple[] {
  if (asset.kind !== "mesh") {
    return [];
  }
  const samples = asset.primitive === "custom" ? sampleCustomMeshPoints(asset) : samplePrimitiveMeshPoints(asset);
  const maxSamples = options.maxSamples ?? samples.length;
  return samples.slice(0, Math.max(0, maxSamples));
}

export function meshAabb(asset: IAssetIr): IAabb | undefined {
  const samples = sampleMeshPoints(asset);
  if (samples.length === 0) {
    return undefined;
  }
  const first = samples[0] as Vec3Tuple;
  const min: [number, number, number] = [first[0], first[1], first[2]];
  const max: [number, number, number] = [first[0], first[1], first[2]];
  for (const sample of samples.slice(1)) {
    min[0] = Math.min(min[0], sample[0]);
    min[1] = Math.min(min[1], sample[1]);
    min[2] = Math.min(min[2], sample[2]);
    max[0] = Math.max(max[0], sample[0]);
    max[1] = Math.max(max[1], sample[1]);
    max[2] = Math.max(max[2], sample[2]);
  }
  return { max, min };
}

export function meshBoundingSphere(asset: IAssetIr): IBoundingSphere | undefined {
  const bounds = meshAabb(asset);
  if (bounds === undefined) {
    return undefined;
  }
  const center: [number, number, number] = [
    (bounds.min[0] + bounds.max[0]) / 2,
    (bounds.min[1] + bounds.max[1]) / 2,
    (bounds.min[2] + bounds.max[2]) / 2,
  ];
  const radius = Math.max(...sampleMeshPoints(asset).map((point) => distance(point, center)));
  return { center, radius };
}

export function aabbIntersectsAabb(left: IAabb, right: IAabb): boolean {
  return left.min[0] <= right.max[0] && left.max[0] >= right.min[0] && left.min[1] <= right.max[1] && left.max[1] >= right.min[1] && left.min[2] <= right.max[2] && left.max[2] >= right.min[2];
}

export function sphereIntersectsSphere(left: IBoundingSphere, right: IBoundingSphere): boolean {
  return distance(left.center, right.center) <= left.radius + right.radius;
}

function sampleCustomMeshPoints(asset: Extract<IAssetIr, { kind: "mesh" }>): Vec3Tuple[] {
  const position = asset.attributes?.find((attribute) => attribute.name === "position" && attribute.itemSize === 3);
  if (position === undefined) {
    return [];
  }
  const samples: Vec3Tuple[] = [];
  for (let index = 0; index < position.values.length; index += 3) {
    samples.push([position.values[index] ?? 0, position.values[index + 1] ?? 0, position.values[index + 2] ?? 0]);
  }
  return samples;
}

function samplePrimitiveMeshPoints(asset: Extract<IAssetIr, { kind: "mesh" }>): Vec3Tuple[] {
  const size = asset.size ?? [];
  if (asset.primitive === "box" || asset.primitive === "extrudedRectangle") {
    const halfX = (size[0] ?? 1) / 2;
    const halfY = (size[1] ?? 1) / 2;
    const halfZ = (size[2] ?? 1) / 2;
    return [
      [-halfX, -halfY, -halfZ],
      [halfX, halfY, halfZ],
    ];
  }
  if (asset.primitive === "plane") {
    const halfX = (size[0] ?? 1) / 2;
    const halfY = (size[1] ?? 1) / 2;
    return [
      [-halfX, -halfY, 0],
      [halfX, halfY, 0],
    ];
  }
  const radius = primitiveRadius(asset);
  const halfHeight = primitiveHalfHeight(asset);
  return [
    [-radius, -halfHeight, -radius],
    [radius, halfHeight, radius],
  ];
}

function primitiveRadius(asset: Extract<IAssetIr, { kind: "mesh" }>): number {
  if (asset.primitive === "torus" || asset.primitive === "annulus") {
    return asset.size?.[1] ?? 1;
  }
  if (asset.primitive === "conicalFrustum") {
    return Math.max(asset.size?.[0] ?? 0.25, asset.size?.[1] ?? 0.5);
  }
  if (asset.primitive === "regularPolygon" || asset.primitive === "circle" || asset.primitive === "sphere") {
    return asset.size?.[0] ?? 0.5;
  }
  return asset.size?.[0] ?? 0.5;
}

function primitiveHalfHeight(asset: Extract<IAssetIr, { kind: "mesh" }>): number {
  if (asset.primitive === "sphere" || asset.primitive === "torus") {
    return primitiveRadius(asset);
  }
  if (asset.primitive === "annulus" || asset.primitive === "circle" || asset.primitive === "regularPolygon") {
    return 0;
  }
  return (asset.size?.[1] ?? 1) / 2;
}

function distance(left: Vec3Tuple, right: Vec3Tuple): number {
  return Math.hypot(left[0] - right[0], left[1] - right[1], left[2] - right[2]);
}

File: runtime-bevy/crates/threenative_runtime/src/mesh_bounds.rs
use threenative_loader::AssetIr;

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct AabbBounds {
    pub min: [f32; 3],
    pub max: [f32; 3],
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct BoundingSphere {
    pub center: [f32; 3],
    pub radius: f32,
}

pub fn sample_mesh_points(asset: &AssetIr, max_samples: Option<usize>) -> Vec<[f32; 3]> {
    if asset.kind != "mesh" {
        return Vec::new();
    }
    let samples = if asset.primitive.as_deref() == Some("custom") {
        sample_custom_mesh_points(asset)
    } else {
        sample_primitive_mesh_points(asset)
    };
    samples
        .into_iter()
        .take(max_samples.unwrap_or(usize::MAX))
        .collect()
}

pub fn mesh_aabb(asset: &AssetIr) -> Option<AabbBounds> {
    let samples = sample_mesh_points(asset, None);
    let first = *samples.first()?;
    let mut min = first;
    let mut max = first;
    for sample in samples.iter().skip(1) {
        for index in 0..3 {
            min[index] = min[index].min(sample[index]);
            max[index] = max[index].max(sample[index]);
        }
    }
    Some(AabbBounds { min, max })
}

pub fn mesh_bounding_sphere(asset: &AssetIr) -> Option<BoundingSphere> {
    let bounds = mesh_aabb(asset)?;
    let center = [
        (bounds.min[0] + bounds.max[0]) / 2.0,
        (bounds.min[1] + bounds.max[1]) / 2.0,
        (bounds.min[2] + bounds.max[2]) / 2.0,
    ];
    let radius = sample_mesh_points(asset, None)
        .iter()
        .map(|point| distance(*point, center))
        .fold(0.0, f32::max);
    Some(BoundingSphere { center, radius })
}

pub fn aabb_intersects_aabb(left: AabbBounds, right: AabbBounds) -> bool {
    left.min[0] <= right.max[0]
        && left.max[0] >= right.min[0]
        && left.min[1] <= right.max[1]
        && left.max[1] >= right.min[1]
        && left.min[2] <= right.max[2]
        && left.max[2] >= right.min[2]
}

pub fn sphere_intersects_sphere(left: BoundingSphere, right: BoundingSphere) -> bool {
    distance(left.center, right.center) <= left.radius + right.radius
}

fn sample_custom_mesh_points(asset: &AssetIr) -> Vec<[f32; 3]> {
    let Some(position) = asset.attributes.as_deref().and_then(|attributes| {
        attributes
            .iter()
            .find(|attribute| attribute.name == "position")
    }) else {
        return Vec::new();
    };
    position
        .values
        .chunks_exact(3)
        .map(|chunk| [chunk[0], chunk[1], chunk[2]])
        .collect()
}

fn sample_primitive_mesh_points(asset: &AssetIr) -> Vec<[f32; 3]> {
    let size = asset.size.as_deref().unwrap_or(&[]);
    match asset.primitive.as_deref() {
        Some("box") | Some("extrudedRectangle") => {
            let half_x = size.first().copied().unwrap_or(1.0) / 2.0;
            let half_y = size.get(1).copied().unwrap_or(1.0) / 2.0;
            let half_z = size.get(2).copied().unwrap_or(1.0) / 2.0;
            vec![[-half_x, -half_y, -half_z], [half_x, half_y, half_z]]
        }
        Some("plane") => {
            let half_x = size.first().copied().unwrap_or(1.0) / 2.0;
            let half_y = size.get(1).copied().unwrap_or(1.0) / 2.0;
            vec![[-half_x, -half_y, 0.0], [half_x, half_y, 0.0]]
        }
        _ => {
            let radius = primitive_radius(asset);
            let half_height = primitive_half_height(asset);
            vec![
                [-radius, -half_height, -radius],
                [radius, half_height, radius],
            ]
        }
    }
}

fn primitive_radius(asset: &AssetIr) -> f32 {
    let size = asset.size.as_deref().unwrap_or(&[]);
    match asset.primitive.as_deref() {
        Some("torus") | Some("annulus") => size.get(1).copied().unwrap_or(1.0),
        Some("conicalFrustum") => size
            .first()
            .copied()
            .unwrap_or(0.25)
            .max(size.get(1).copied().unwrap_or(0.5)),
        Some("regularPolygon") | Some("circle") | Some("sphere") => {
            size.first().copied().unwrap_or(0.5)
        }
        _ => size.first().copied().unwrap_or(0.5),
    }
}

fn primitive_half_height(asset: &AssetIr) -> f32 {
    match asset.primitive.as_deref() {
        Some("sphere") | Some("torus") => primitive_radius(asset),
        Some("annulus") | Some("circle") | Some("regularPolygon") => 0.0,
        _ => {
            asset
                .size
                .as_deref()
                .and_then(|size| size.get(1))
                .copied()
                .unwrap_or(1.0)
                / 2.0
        }
    }
}

fn distance(left: [f32; 3], right: [f32; 3]) -> f32 {
    ((left[0] - right[0]).powi(2) + (left[1] - right[1]).powi(2) + (left[2] - right[2]).powi(2))
        .sqrt()
}
